Update buildchain and CI for operator

slaperche-scality opened the pull request:
**Component**:

ci, buildchain

**Context**: 

In order to deploy the operator, we need to be able to build it and add it in the ISO.

**Summary**:

Add new dependencies in the CI to build the operator (`go`, `operator-sdk` and `dep`).
Update the buildchain with new tasks:
- to build the operator image and add it to the ISO
- to lint the Go code (and check that the generated files are up-to-date)
- to help formatting Go code (using `gofmt`) and generating boilerplate code (using `operator-sdk generate`).

**Acceptance criteria**: 

When building the ISO, you should have the operator image in the build tree (check with `ls _build/root/images/storage-operator/*`)
If you're familiar with Go, you can also edit `storage-operator/pkg/apis/storage/v1alpha1/volume_types.go` to test the new tasks:
- `./doit.sh format` should reformat the code according to `gofmt`
- `./doit.sh codegen` should re-generate the boilerplate code
- `./doit.sh lint:go` should complains if either the formatting is not good or the generated files are outdated.

---

Closes: #1410

slaperche-scality committed:
doc/build: update the requirements for operators

Also add a minimum required version for Docker (according to the README
of the Operator SDK)

Refs: #1410
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
ci/worker: add operator-sdk to the build container

We now need to build the Storage Operator, which we built using the
Operator SDK.
Consequently, we need to add the Operator SDK to the build image.

Since the SDK depends on Go and dep, we also install those.

Refs: #1410
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
build/image: split the build and save step

That way, we will be able to selectively override one while reusing the
other (e.g. keep saving with Skopeo, but build the image using the
operator-sdk instead of Docker directly).

Refs: #1410
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
build/deps: add a property for the command name

Until now, the command-name was simply the name of the enum variant in
lowercase, but this will no longer be the case with
OPERATOR_SDK/operator-sdk.

Refs: #1410
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
build/deps: workaround a pylint bug

Pylint seems to think that `name` is a method instead of a string, which
is wrong.
See https://github.com/PyCQA/pylint/issues/2062

Refs: #1410
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
build/image: build the storage-operator image

Add a new image to be built: the container image for the Storage
operator.
Since it's very similar to a LocalImage, except for the building part
(we use the operator-sdk command instead of Docker directly, because we
need to compile the Go code), we introduce a new class (OperatorImage)
that inherits from LocalImage and overload the `_do_build` method.

Note that we take care of enabling Go module and compiling from the
`storage-operator` directory.

Refs: #1410
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
build/lint: add a task to lint the Go code

For now, we only check that the code is properly formated using `gofmt`.

Refs: #1410
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
build/lint: add title for lint tasks

Until now, lint tasks weren't run alongside "build" task, and thus the
discrepancy in the display wasn't a bother.
But now, we have this kind of output, which is unsightly:

    -- lint:python
    -- lint:shell
    -- lint:yaml
    .  CHECK CMD    gofmt
    -- lint:go

So let's define a custom title for the lint tasks to have something like
that instead:

    -- LINT         python
    -- LINT         shell
    -- LINT         yaml
    .  CHECK CMD    gofmt
    -- LINT         go

Refs: #1410
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
build/lint: update outdated docstring

We added new tasks but the schema wasn't updated.

Refs: #1410
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
build/format: add a task to auto-format the code

For now, we only have a task for Go code.
If we ever use `black` for Python, we will add a new subtask here.

Refs: #1410
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
build: compute gofmt args only once

The arguments list was computed twice: once in `lint` and once in
`format`.
Let's use a constant instead.

Refs: #1410
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
build: add an helper for title with sub-task name

This reduce code duplication between `lint` and `format`.

Refs: #1410
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
build/utils: fix a typo in the docstring

Refs: #1410
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
build/lint: add an action to check generated file

We commit the files generated by the `operator-sdk` command, as such we
want to make sure that when modifying the "source" file we don't forget
to regenerate the scaffolding files.

Refs: #1410
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
build/codegen: add a task to run Go codegen

For now, I plan to create one task per language, for the sake of
consistency with `lint` and `format`.
If we have more operator in the future, it may make more sense to have
one subtask per operator: we can change when/if it happens.

Refs: #1410
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

bert-e commented:
# Hello slaperche-scality,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @nootal

* @cmonfort

* @lucieleonard

* @alxf

* @carlito-scality

* @alexandre-allard-scality

* @teddyandrieux

* @slaperche-scality

* @thomasdanan

* @zempashi

* @nicolast

* @sayf-eddine-scality

* @monpote

* @gdemonet

* @tcarmet

slaperche-scality commented:
Need to be sure it works on the CI (i.e.: green build) before removing the Draft label.
Also want to check if `dep` is really required, it may not be.

Other than that, you can start reviewing the code (lot of commit, but several of them are pretty small).

NicolasT commented:
We don't (or shouldn't, at least) rely on `dep`, we're using Go modules and no vendoring.

NicolasT commented on buildchain/buildchain/config.py:
```diff
@@ -52,10 +52,21 @@
+    SKOPEO       = os.getenv('SKOPEO_BIN',       'skopeo')
+    VAGRANT      = os.getenv('VAGRANT_BIN',      'vagrant')
+    OPERATOR_SDK = os.getenv('OPERATOR_SDK_BIN', 'operator-sdk')
+    GOFMT        = os.getenv('GOFMT_BIN',        'gofmt')
```
`go fmt` ?

slaperche-scality replied:
I tried using `go fmt` but it only supports a subset of what `goftmt` does (no `-s` and more importantly, no `-d`)

```
% go help fmt
usage: go fmt [-n] [-x] [packages]

Fmt runs the command 'gofmt -l -w' on the packages named
by the import paths. It prints the names of the files that are modified.

For more about gofmt, see 'go doc cmd/gofmt'.
For more about specifying packages, see 'go help packages'.

The -n flag prints commands that would be executed.
The -x flag prints commands as they are executed.

To run gofmt with specific options, run gofmt itself.

See also: go fix, go vet.
% gofmt -h
usage: gofmt [flags] [path ...]
  -cpuprofile string
        write cpu profile to this file
  -d    display diffs instead of rewriting files
  -e    report all errors (not just the first 10 on different lines)
  -l    list files whose formatting differs from gofmt's
  -r string
        rewrite rule (e.g., 'a[b:len(a)] -> a[b:]')
  -s    simplify code
  -w    write result to (source) file instead of stdout
```
See also: https://stackoverflow.com/questions/55319495/what-is-the-difference-between-gofmt-go-fmt

Since both are available with a standard Go installation (`gofmt` isn't an extra-dependency), I went with `gofmt`.

NicolasT commented on buildchain/buildchain/constants.py:
```diff
@@ -103,3 +105,13 @@ def git_ref() -> Optional[str]:
+# Only keep directories (except `vendor`) and top-level Go source files.
+STORAGE_OPERATOR_FMT_ARGS : FrozenSet[str] = frozenset([
+    path.name for path in STORAGE_OPERATOR_ROOT.glob('*')
+    if (path.is_dir() and path.name != 'vendor') or path.suffix == '.go'
```
There should be no `vendor` directory. If there is one, that's a 'bug'.

slaperche-scality replied:
Ho, you're right!!
The one I had is a vestigial directory of when I was using an old `operator-sdk`.

I guess we could also remove the entry from the `.gitignore` then?

slaperche-scality replied:
Done

NicolasT commented on eve/workers/pod-builder/Dockerfile:
```diff
@@ -26,4 +26,20 @@ RUN yum install -y epel-release \
 
+# Install Go tooling to build the Kubernetes Operators
+ENV GO_VERSION 1.12.7
+ENV OP_SDK_VERSION v0.9.0
```
I'd avoid using shorthands.

slaperche-scality replied:
Done

NicolasT commented on eve/workers/pod-builder/Dockerfile:
```diff
@@ -26,4 +26,20 @@ RUN yum install -y epel-release \
+    && mv go /usr/local                                                        \
+    && curl -RLo /usr/bin/operator-sdk                                         \
+        https://github.com/operator-framework/operator-sdk/releases/download/${OP_SDK_VERSION}/operator-sdk-${OP_SDK_VERSION}-x86_64-linux-gnu \
+    && chmod +x /usr/bin/operator-sdk
```
Consider checking the signature of the downloaded file?

slaperche-scality replied:
:+1: 
I'll add it (anyway I have to modify this Dockerfile, things aren't working yet)

slaperche-scality replied:
That may be harder than expected, because I don't have the maintainers public key, so I have to download it but to get the ID I need to parse the error message from `gpg`

> To download the key, use the following command, replacing $KEY_ID with the RSA key string provided in the output of the previous command:

:/

slaperche-scality replied:
Or should I hardcode the key when doing the `recv-keys`?

NicolasT replied:
Ugh, maybe file a TODO ticket but leave it for now.

slaperche-scality replied:
Tracked by #1444.

slaperche-scality committed:
ci/worker: add operator-sdk to the lint container

Since the lint step for Go uses `gofmt` and `operator-sdk`, we also need
to install them in the lint worker.

Refs: #1410
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
build: remove `vendor` filtering

Since we're using Operator SDK 0.9.0, we no longer have a `vendor`
directory, so there is no need to filter it and its content.

Refs: #1410
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality commented:
> We don't (or shouldn't, at least) rely on `dep`, we're using Go modules and no vendoring.

Yes, you're right: I got rid of it.

slaperche-scality commented:
A green build \o/, I'm passing this as ready for review then :slightly_smiling_face: 

That being said, since it's adding 20 min to the CI time (because of the retries) we probably want to fix that.
But "it works" so it can be reviewed…

NicolasT commented:
Adding 20min to CI is a no-go as far as I'm concerned. I believe there's a fairly easy fix though: we're running the build in a container, right? (If not, shouldn't we? ;-))

We can then apply the same 'trick' as we do for the doc building which downloads the Tox/Sphinx env, something along the lines of
```
WORKDIR /usr/src/storage-operator
COPY go.mod go.sum /usr/src/storage-operator/
RUN go mod download
COPY . /usr/src/storage-operator
RUN operator-sdk build ...
```

NicolasT commented on storage-operator/.gitignore:
```diff
@@ -1,5 +1,3 @@
-vendor/
```
I wouldn't remove it, TBH. *Not* having it wouldn't prevent anyone from checking things into `vendor/`. Having it requires one to add `-f` to `git add`, otherwise Git won't be willing to touch files in that folder.

slaperche-scality replied:
But how do they check things into `vendor`? The directory no longer exists when using 0.9.0.

NicolasT replied:
`mkdir vendor`? `go mod vendor`?...

Boils down to: having `vendor/` in here doesn't do any harm, it might even provide a benefit, so I don't see any reason to remove it.

slaperche-scality replied:
Fair enough, I'll add it back then.

But if we could have `vendor` here from `go mod vendor` I should keep the code to filter it then, no?

NicolasT replied:
No, because we *shouldn't* have a `vendor` directory. This is a way to ensure we don't ever get one (unless one is very stubborn).

slaperche-scality replied:
Ok seen with you directly: will keep the `.gitignore` entry but remove the code

slaperche-scality committed:
ci/worker: pre-download Go dependencies

We need to download a fairly big amount of data when building operator
using the Operator SDK (~3G of source repositories).

In order to avoid stressing the CI (and wasting bandwidth) by repeatedly
doing this at build time for every single build, let's pre-download all
that stuff and keep it in the container image.

Note that since we need to copy the `go.sum` and `go.mod` files, we need
to set the build context in the storage-operator directory.

The `pod-linter` is also affected as we use the `operator-sdk` command
during our linting phase.

Closes: #1410
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/1410-update-buildchain-and-ci-for-operator` into
`development/2.2`
* [w/2.3/feature/1410-update-buildchain-and-ci-for-operator](https://github.com/scality/metalk8s/compare/development/2.3...w/2.3/feature/1410-update-buildchain-and-ci-for-operator) will be merged into `development/2.3` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @nootal

* @cmonfort

* @lucieleonard

* @alxf

* @carlito-scality

* @alexandre-allard-scality

* @teddyandrieux

* @slaperche-scality

* @thomasdanan

* @zempashi

* @nicolast

* @sayf-eddine-scality

* @monpote

* @gdemonet

* @tcarmet

bert-e commented:
# History mismatch



Merge commit #917f4fb4eee92c3e5f1d4133f788a75f503afd10 on the integration branch
`w/2.3/feature/1410-update-buildchain-and-ci-for-operator` is merging a branch which is neither the current
branch `feature/1410-update-buildchain-and-ci-for-operator` nor the development branch
`development/2.3`.

It is likely due to a rebase of the branch `feature/1410-update-buildchain-and-ci-for-operator` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

bert-e commented:
# History mismatch



Merge commit #55acd26703493e3a09a543a421c06c645a86c050 on the integration branch
`w/2.3/feature/1410-update-buildchain-and-ci-for-operator` is merging a branch which is neither the current
branch `feature/1410-update-buildchain-and-ci-for-operator` nor the development branch
`development/2.3`.

It is likely due to a rebase of the branch `feature/1410-update-buildchain-and-ci-for-operator` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

slaperche-scality commented:
@bert-e reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/1410-update-buildchain-and-ci-for-operator` into
`development/2.2`
* [w/2.3/feature/1410-update-buildchain-and-ci-for-operator](https://github.com/scality/metalk8s/compare/development/2.3...w/2.3/feature/1410-update-buildchain-and-ci-for-operator) will be merged into `development/2.3` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @nootal

* @cmonfort

* @lucieleonard

* @alxf

* @carlito-scality

* @alexandre-allard-scality

* @teddyandrieux

* @slaperche-scality

* @thomasdanan

* @zempashi

* @nicolast

* @sayf-eddine-scality

* @monpote

* @gdemonet

* @tcarmet

bert-e commented:
# History mismatch



Merge commit #bcc59b76d8e5ae65df41ded431adec11618c5cd3 on the integration branch
`w/2.3/feature/1410-update-buildchain-and-ci-for-operator` is merging a branch which is neither the current
branch `feature/1410-update-buildchain-and-ci-for-operator` nor the development branch
`development/2.3`.

It is likely due to a rebase of the branch `feature/1410-update-buildchain-and-ci-for-operator` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

NicolasT commented on eve/workers/pod-linter/Dockerfile:
```diff
@@ -20,8 +21,28 @@ RUN dnf install -y git \
   && pip install tox \
   && pip install buildbot-worker==${BUILDBOT_VERSION}
 
+# Install Go tooling to lint the Kubernetes Operators.
```
Given we seem to be creating (almost) the same container image twice, does it really make sense to make a distinction between a `linter` and `builder` container?

slaperche-scality replied:
I think the reason behind the split is that we want to use `centos:7` as base image (and that's what we do for the pod-builder).

Now, the issue with pod-linter is that we need a not too old `shellcheck`, which is not available on CentOS 7 so Jordi switched the base image to `fedora:29` for this pod (IIRC it was too cumbersome to get a recent `shellcheck` on CentOS)

That's the reason behind the split I think (keep using CentOS 7 everywhere, except for linting).

bert-e commented:
# History mismatch



Merge commit #7bd016059a61c17c5648d934017fda510557380e on the integration branch
`w/2.3/feature/1410-update-buildchain-and-ci-for-operator` is merging a branch which is neither the current
branch `feature/1410-update-buildchain-and-ci-for-operator` nor the development branch
`development/2.3`.

It is likely due to a rebase of the branch `feature/1410-update-buildchain-and-ci-for-operator` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

slaperche-scality commented:
@bert-e reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/1410-update-buildchain-and-ci-for-operator` into
`development/2.2`
* [w/2.3/feature/1410-update-buildchain-and-ci-for-operator](https://github.com/scality/metalk8s/compare/development/2.3...w/2.3/feature/1410-update-buildchain-and-ci-for-operator) will be merged into `development/2.3` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @nootal

* @cmonfort

* @lucieleonard

* @alxf

* @carlito-scality

* @alexandre-allard-scality

* @teddyandrieux

* @slaperche-scality

* @thomasdanan

* @zempashi

* @nicolast

* @sayf-eddine-scality

* @monpote

* @gdemonet

* @tcarmet

bert-e commented:
# History mismatch



Merge commit #824d2d8f9d44b1314797c793aae3e74aede81b5b on the integration branch
`w/2.3/feature/1410-update-buildchain-and-ci-for-operator` is merging a branch which is neither the current
branch `feature/1410-update-buildchain-and-ci-for-operator` nor the development branch
`development/2.3`.

It is likely due to a rebase of the branch `feature/1410-update-buildchain-and-ci-for-operator` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

slaperche-scality commented:
@bert-e reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/1410-update-buildchain-and-ci-for-operator` into
`development/2.2`
* [w/2.3/feature/1410-update-buildchain-and-ci-for-operator](https://github.com/scality/metalk8s/compare/development/2.3...w/2.3/feature/1410-update-buildchain-and-ci-for-operator) will be merged into `development/2.3` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @nootal

* @cmonfort

* @lucieleonard

* @alxf

* @carlito-scality

* @alexandre-allard-scality

* @teddyandrieux

* @slaperche-scality

* @thomasdanan

* @zempashi

* @nicolast

* @sayf-eddine-scality

* @monpote

* @gdemonet

* @tcarmet

bert-e commented:
# History mismatch



Merge commit #2b896d9dac61bd5a9616671e1253015621affec0 on the integration branch
`w/2.3/feature/1410-update-buildchain-and-ci-for-operator` is merging a branch which is neither the current
branch `feature/1410-update-buildchain-and-ci-for-operator` nor the development branch
`development/2.3`.

It is likely due to a rebase of the branch `feature/1410-update-buildchain-and-ci-for-operator` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

gdemonet commented on buildchain/buildchain/config.py:
```diff
@@ -52,10 +52,21 @@
+    @property
+    def command_name(self) -> str:
+        """Return the name of the command."""
+        if self is self.OPERATOR_SDK:
```
Mmh, works but could have been simpler (any command name containing dashes will need this special case). On the other hand, the heuristic is quite simple:

```python
def command_name(self):
    return self.name.lower().replace('_', '-')
```

slaperche-scality replied:
The problem is that some command does contains `_` in their name and that would break them.

True, we aren't using such command yet, but if one day we add one, I don't want the poor guy to go looking all over the buildchain trying to understand why the command he added doesn't work.

gdemonet commented on buildchain/buildchain/image.py:
```diff
@@ -344,6 +344,13 @@ def show() -> str:
     ),
+    targets.OperatorImage(
+        name='storage-operator',
+        version='latest',
```
Don't we want the MetalK8s version here?

slaperche-scality replied:
In the long term, we do want a real version number here (not latest).
For now: no, we will stay with latest.

Extract from a Slack discussion I had with Nicolas:

> Can use `latest` for now
> Until we figure out how to properly version `storage-operator`, also in its `version.go` file and such
> Can of worms

gdemonet commented on eve/workers/pod-builder/Dockerfile:
```diff
@@ -19,11 +19,32 @@ RUN yum install -y epel-release \
 
+# Install Go tooling to build the Kubernetes Operators.
+ENV GO_VERSION 1.12.7
+ENV OPERATOR_SDK_VERSION v0.9.0
```
Shouldn't these versions be set in `ARG` commands instead? Same as for `BUILDBOT_VERSION`

slaperche-scality replied:
Done

gdemonet commented on buildchain/buildchain/format.py:
```diff
@@ -0,0 +1,48 @@
+    }
+
+
+# List of available formatting task.
```
minor:

```suggestion
# List of available formatting tasks.
```

slaperche-scality replied:
Done (also fixed the same in `codegenpy`)

gdemonet commented on buildchain/buildchain/utils.py:
```diff
@@ -61,3 +61,19 @@ def title_with_target1(command: str, task: types.Task) -> str:
     )
+
+
+def title_with_subtask_name(command: str, task: types.Task) -> str:
```
minor: Could be interesting to have these `title_with_<something>` wrapped in another function, taking only the `command` argument, since we'll always need a `function(task: types.Task) -> str` to pass in the task dict.

slaperche-scality replied:
Good idea!

Done in a dedicated commit.

gdemonet commented on buildchain/buildchain/lint.py:
```diff
@@ -64,27 +75,86 @@ def lint_shell() -> types.TaskDict:
+    # the generated
+    if current != base:
+        return doit.exceptions.TaskError(
+            msg='outdated generated Go files, did you run `doit.sh codegen:go`?'
```
minor: for consistency with `check_go_fmt`

```suggestion
            msg='outdated generated Go files, please run `doit.sh codegen:go`'
```

slaperche-scality replied:
Since the lint does run the codegen, the user don't really have to run it: we have done it for him already (would be better if `operator-sdk` had a dry-run mode, but that's not the case, so…)

gdemonet commented on buildchain/buildchain/lint.py:
```diff
@@ -64,27 +75,86 @@ def lint_shell() -> types.TaskDict:
+        subprocess.check_call(cmd, cwd=cwd)
+    current = subprocess.check_output(git_diff)
+    # If the diff changed after running the code generation that means that
+    # the generated
```
This comment is missing an end I believe ;)

slaperche-scality replied:
Indeed >_<

Done

gdemonet commented on buildchain/buildchain/lint.py:
```diff
@@ -64,27 +75,86 @@ def lint_shell() -> types.TaskDict:
+    current = subprocess.check_output(git_diff)
+    # If the diff changed after running the code generation that means that
+    # the generated
+    if current != base:
```
Should we reset the changes in that situation?

slaperche-scality replied:
If we reset, we could lose the existing diff of the user.

The buildchain should avoid to mess with the git state as much as possible (yes, here we are modifying files, but anyway the guy is gonna run `./doit.sh codegen` so it's not a big deal).

gdemonet commented on eve/workers/pod-builder/Dockerfile:
```diff
@@ -19,11 +19,32 @@ RUN yum install -y epel-release \
 USER eve
+
+# Pre-download the Go dependencies.
+COPY storage-operator/go.mod storage-operator/go.sum ./
```
Since you're only `COPY`ing from `storage-operator/`, maybe the build context could be just `'storage-operator/'` instead of `'.'`?

slaperche-scality replied:
Indeed. Done.

gdemonet reviewed (commented):
Few questions and minor comments, most of it looks good to me!

slaperche-scality committed:
build/utils: return a closure from title_with_*

That way, the callers don't need to use a lambda or an extra function.
This make the code simpler.

Refs: #1410
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

bert-e commented:
# History mismatch



Merge commit #eb4553f39c7f051430b4fae28eb434a4e1fd5827 on the integration branch
`w/2.3/feature/1410-update-buildchain-and-ci-for-operator` is merging a branch which is neither the current
branch `feature/1410-update-buildchain-and-ci-for-operator` nor the development branch
`development/2.3`.

It is likely due to a rebase of the branch `feature/1410-update-buildchain-and-ci-for-operator` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

bert-e commented:
# History mismatch



Merge commit #b3b107f147d4739644a9beadfc2c995da085ad78 on the integration branch
`w/2.3/feature/1410-update-buildchain-and-ci-for-operator` is merging a branch which is neither the current
branch `feature/1410-update-buildchain-and-ci-for-operator` nor the development branch
`development/2.3`.

It is likely due to a rebase of the branch `feature/1410-update-buildchain-and-ci-for-operator` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

slaperche-scality commented:
@bert-e reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/1410-update-buildchain-and-ci-for-operator` into
`development/2.2`
* [w/2.3/feature/1410-update-buildchain-and-ci-for-operator](https://github.com/scality/metalk8s/compare/development/2.3...w/2.3/feature/1410-update-buildchain-and-ci-for-operator) will be merged into `development/2.3` 
* [w/2.4/feature/1410-update-buildchain-and-ci-for-operator](https://github.com/scality/metalk8s/compare/development/2.4...w/2.4/feature/1410-update-buildchain-and-ci-for-operator) will be merged into `development/2.4` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @nootal

* @cmonfort

* @lucieleonard

* @alxf

* @carlito-scality

* @alexandre-allard-scality

* @teddyandrieux

* @slaperche-scality

* @thomasdanan

* @zempashi

* @nicolast

* @sayf-eddine-scality

* @monpote

* @gdemonet

* @tcarmet

NicolasT commented:
It would make sense to hold off on this PR until #1447 is merged, then rebase it on `development/2.4` and change the target branch, so other version branches aren't impacted by the time spent on this in CI.

bert-e commented:
# History mismatch



Merge commit #dfa8e5998d34b6ed18e1c9aff2d13cb02d2cd1cf on the integration branch
`w/2.3/feature/1410-update-buildchain-and-ci-for-operator` is merging a branch which is neither the current
branch `feature/1410-update-buildchain-and-ci-for-operator` nor the development branch
`development/2.3`.

It is likely due to a rebase of the branch `feature/1410-update-buildchain-and-ci-for-operator` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

bert-e commented:
# History mismatch



Merge commit #4721f71d05e8eadc477673789053cd2ff7242fa2 on the integration branch
`w/2.3/feature/1410-update-buildchain-and-ci-for-operator` is merging a branch which is neither the current
branch `feature/1410-update-buildchain-and-ci-for-operator` nor the development branch
`development/2.3`.

It is likely due to a rebase of the branch `feature/1410-update-buildchain-and-ci-for-operator` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @nootal

* @cmonfort

* @lucieleonard

* @alxf

* @carlito-scality

* @alexandre-allard-scality

* @teddyandrieux

* @slaperche-scality

* @thomasdanan

* @zempashi

* @nicolast

* @sayf-eddine-scality

* @monpote

* @gdemonet

* @tcarmet

slaperche-scality commented:
@bert-e reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @nootal

* @cmonfort

* @lucieleonard

* @alxf

* @carlito-scality

* @alexandre-allard-scality

* @teddyandrieux

* @slaperche-scality

* @thomasdanan

* @zempashi

* @nicolast

* @sayf-eddine-scality

* @monpote

* @gdemonet

* @tcarmet

NicolasT commented on buildchain/buildchain/codegen.py:
```diff
@@ -0,0 +1,49 @@
+    """Generate Go code using operator-sdk."""
+    cwd  = constants.STORAGE_OPERATOR_ROOT
+    actions = []
+    for target in ('k8s', 'openapi'):
```
To be really picky, this should be `['k8s', 'openapi']` (since it's a `list<str>` rather than a `(str, str)`), but let's not let that keep this PR unmerged :-P

gdemonet replied:
Why? Tuples are lighter than lists

slaperche-scality replied:
I used tuple for its immutability (since the list is kind of a constant here), but in this context tuple and list are really alike: won't make any (observable) difference.

In a statically typed language, tuples and lists are a different beasts (lists are homogeneous whereas tuples aren't), but in Python where you can mix apples and oranges in both lists and tuples the key difference is the mutability.

NicolasT replied:
The remark came indeed from a 'static typing' background:

- A *list* is an *ordered* collection of an *unknown* number of values of a *single* type
- A *tuple* is an *ordered* collection of a *known* number of values of (potentially) *different* (but known) types

As such, the type constructor of a list has a single argument, whilst the type constructor of an N-tuple has N arguments.

The immutability of a tuple doesn't really play a role in the argument.

Anyway, pedantic, though in general I'm not overly fond of conflating the two (yes, despite duck-typing).

NicolasT commented on storage-operator/.dockerignore:
```diff
@@ -0,0 +1,5 @@
+cmd/
```
Not very fond we need this, but sure, for now. At some point I'd like to roll building of the whole thing in a container-based environment (indeed, maybe not using `operator-sdk build ...`), at which point this would need to be removed. Someday.

slaperche-scality replied:
Yes I added it to only upload the strict minimum to the docker daemon.

But that looks like premature optimization (and I even broke `operator-sdk build` when I added `build/` to the ignore => proves that it wasn't a great idea ^^").

I won't be against its removal when it becomes an hindrance :slightly_smiling_face:

slaperche-scality commented:
@bert-e approve

slaperche-scality commented:
@bert-e approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.4`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.4`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



Please check the status of the associated issue None.

Goodbye slaperche-scality.
